nemo_curator/utils/distributed_utils.py

```diff
@@ -343,13 +360,79 @@ def read_single_partition(
     else:
         df = read_f(files, **read_kwargs, **kwargs)
 
-    if filetype in ["jsonl", "json"] and columns is not None:
-        if add_filename and "filename" not in columns:
-            columns.append("filename")
-        df = df[columns]
+    print(f"Reading with {read_kwargs=}", flush=True)
+    return select_and_sort_columns(df, columns, add_filename)
 
-    df = df[sorted(df.columns)]
-    return df
+
+def create_file_groups(
+    input_files: List[str],
+    files_per_partition: Optional[int],
+    blocksize: Optional[str],
+) -> List[List[str]]:
+    if files_per_partition is not None:
+        # When FPP is specified we just sort, and create buckets of FPP size each
+        input_files = sorted(input_files)
+        if files_per_partition > 1:
+            input_files = [
+                input_files[i : i + files_per_partition]
+                for i in range(0, len(input_files), files_per_partition)
+            ]
+        else:
+            input_files = [[file] for file in input_files]
+    elif blocksize is not None:
+        # We implement bucketing using First-fit-decreasing bin packing algorithm
+        # This means we sort the files by size, and then try to fit them into buckets
+        # For each file we try all the buckets to see if it fits, if it does we add it to that bucket
+        # If it doesn't fit in any bucket we create a new bucket
+        # The time complexity is O(N*M) where N is the number of files and M is the number of buckets
+
+        from dask.utils import parse_bytes
+
+        blocksize = parse_bytes(blocksize)
+
+        input_size_file_name = [(os.path.getsize(f), f) for f in input_files]
```

Comment:
As you mentioned, I think doing this with `os` would get too heavy for cases with thousands or millions of files. I am in favor of https://github.com/NVIDIA/NeMo-Curator/pull/285.